flextape/service/queue.go

```diff
@@ -0,0 +1,137 @@
+package service
+
+import (
+	"sort"
+)
+
+// invocationq is a simple queue of invocations.
+//
+// Beside implementing the classic Enqueue and Dequeue methods:
+// - It partially implements the sort.Interface to allow the easy
+//   implementation of invocation prioritization policies.
+// - It maintains per-entry state (QueueID) so it's easy to tell
+//   the position of each invocation by just looking at that value.
+type invocationq []*invocation
+
+func (iq *invocationq) Enqueue(x *invocation) Position {
+	position := Position(len(*iq))
+	offset := QueueID(1)
+	if len(*iq) > 0 {
+		offset = (*iq)[0].QueueID
+	}
+
+	x.QueueID = offset + QueueID(position)
+	(*iq) = append(*iq, x)
+	return position
+}
+
+func (iq *invocationq) Dequeue() *invocation {
+	if len(*iq) <= 0 {
+		return nil
```

Comment:
Returning an error here makes the API a bit harder, but reduces the odds of someone forgetting a nil check